### PR TITLE
ci(build): Rust Code Coverage

### PR DESCRIPTION
**Pull request by mtopolnik**
This PR brings Rust test coverage reporting to the QuestDB repo. It transfers the main approach from Questdb Enterprise, and adapts it to the specifics of the OSS build. The most important difference is the need to merge the results of several independent test runs.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 3 / 5 (60.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/Bootstrap.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 5 (20.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 5 (20.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 9 (11.11%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|parquet_write/jni.rs|0|3|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|2|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|2|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|2|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|io/questdb/Bootstrap.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|2|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 4 / 11 (36.36%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|lib.rs|0|1|00.00%|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|io/questdb/jit/JitUtil.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 11 (72.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|3|6|50.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/JitUtil.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 8 (62.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|3|6|50.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 8 (25.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 8 (62.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|3|6|50.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 8 (25.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 8 (25.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 8 (25.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|0|6|00.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

coverage check fail. please retry. :fearful:

[Please let me know](https://github.com/naver/cover-checker/issues/new) when error again.

java.lang.NullPointerException

**Comment by ideoma** [comment]
#### [PR Coverage check]

coverage check fail. please retry. :fearful:

[Please let me know](https://github.com/naver/cover-checker/issues/new) when error again.

java.lang.NullPointerException

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 8 (62.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|3|6|50.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 8 (62.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|3|6|50.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 8 (62.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|3|6|50.00%|
|:large_blue_circle:|lib.rs|1|1|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by ideoma** [comment]
#### [PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: ci/templates/steps.yml [hunk 1]
```diff
@@ -9,14 +9,17 @@ steps:
...
       key: '"questdb_main" | "$(Agent.OS)" | "$(jdk)" | "maven"'
-      restoreKeys: |
-        "questdb_main" | "$(Agent.OS)" | "$(jdk)" | "maven"
       path: $(MAVEN_CACHE_FOLDER)
+  - bash: |
+      echo "SOURCE_CODE_CHANGED: $(SOURCE_CODE_CHANGED)"
+      echo "RUST_SOURCE_CODE_CHANGED: $(RUST_SOURCE_CODE_CHANGED)"
+      echo "COVERAGE_DIFF: $(COVERAGE_DIFF)"
+      echo "CODE_COVERAGE_TOOL_OPTION: $(CODE_COVERAGE_TOOL_OPTION)"
+    displayName: "Print out task variables"
```
Comment (ideoma) [inline ci/templates/steps.yml:18]: nit: you don't need to do this, Azure job log these, you can find them if you expand "Job preparation parameters"


image

File: core/rust/qdbr/Cargo.toml [hunk 2]
```diff
@@ -19,10 +19,11 @@ async-trait = "0.1.81"
 clap = "4.5.9"
 tokio = "1.38.0"
 log = "0.4.22"
+fail = "0.5"
```
Comment (ideoma) [inline core/rust/qdbr/Cargo.toml:22]: nit: is it used?

File: ci/templates/hosted-cover-jobs.yml [hunk 3]
```diff
@@ -8,26 +8,26 @@ jobs:
           poolName: "Azure Pipelines"
           os: Linux
           jdk: "1.11"
-          testset: "coverage"
-          includeTests: "**/griffin/**"
+          testset: "coverage-griffin"
+          includeTests: "**/griffin/engine/table/parquet/**"
```
Comment (ideoma) [inline ci/templates/hosted-cover-jobs.yml:12]: Don't forget to revert this change

File: ci/templates/steps.yml (hunk 1 above)
Comment (mtopolnik) [reply ci/templates/steps.yml:18]: Commit prepared that deletes this.

File: core/rust/qdbr/Cargo.toml (hunk 2 above)
Comment (mtopolnik) [reply core/rust/qdbr/Cargo.toml:22]: Referenced here: https://github.com/questdb/questdb/pull/4830/files#diff-8d04401f1cc51365fe3e32f019cd720135ba920a1a7da7f19e9c9208478701fcR47

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 2 / 3 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|1|2|50.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 2 (50.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|parquet_write/binary.rs|1|2|50.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by ideoma** [comment]
#### [PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)
